Use config.get() for safer access to configuration

DukeDavis12 opened the pull request:
This commit updates direct dictionary access of configuration with safer config.get().This change also prevents potential `KeyError` exceptions if the keys are missing.

Fixes #243, #252, #253 .

<!--
    Please fill in the following form before submitting this PR
    and ensure that your code follows our coding style guideline:
    https://gramine.readthedocs.io/en/latest/devel/coding-style.html -->

## Description of the changes <!-- (reasons and measures) -->

<!--
    If your PR fixes an issue, please remember to add "Fixes #issue_number"
    here, to automatically close it on merge. -->

## How to test this PR? <!-- (if applicable) -->

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/gramineproject/gsc/244)
<!-- Reviewable:end -->

mkow reviewed (changes requested):
> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/244)* status: all files reviewed, 5 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: )

___
*[`-- commits` line 5 at r1](https://reviewable.io/reviews/gramineproject/gsc/244#-OTRPtPbAH5AWteSiXfM:-OTRPtPbAH5AWteSiXfN:b-345iwb):*
nitpick



_Suggestion:_
```
 config.get(). This
```
___
*[`-- commits` line 5 at r1](https://reviewable.io/reviews/gramineproject/gsc/244#-OTRQ--M991aysMcy8Mk:-OTRQ--M991aysMcy8Ml:b-nums7r):*
Also? Isn't that the only thing it prevents?



_Suggestion:_
```
prevents
```
___
*[`gsc.py` line 352 at r1](https://reviewable.io/reviews/gramineproject/gsc/244#-OTRQMjC0RYAr_zjgO_d:-OTRQMjC0RYAr_zjgO_e:b-21uosq) ([raw file](https://github.com/gramineproject/gsc/blob/6660bc9f9eafebdb328306af311012521ee75eae/gsc.py#L352)):*
> ```Python
> 
>     config = yaml.safe_load(args.config_file)
>     if 'Image' in config.get('Gramine'):
> ```

This will still fail if `Gramine` key is not present in the config (you'll get a `TypeError`).
___
*[`gsc.py` line 353 at r1](https://reviewable.io/reviews/gramineproject/gsc/244#-OTRQeOY5bZGo65rMCTs:-OTRQeOY5bZGo65rMCTt:b4q34o8) ([raw file](https://github.com/gramineproject/gsc/blob/6660bc9f9eafebdb328306af311012521ee75eae/gsc.py#L353)):*
> ```Python
>     config = yaml.safe_load(args.config_file)
>     if 'Image' in config.get('Gramine'):
>         gramine_image_name = config.get('Gramine', {}).get('Image')
> ```

This still won't work if the name is missing, `get_docker_image` will get `None` instead of the name and fail.
We either have to provide a default value (which doesn't make sense for this config option, I think), or just error out on the key access (i.e. here) instead of somewhere deeper in the logic. It will be much clearer to the user if they get `"Image" key not found` error, instead of something saying that the Docker daemon got an unexpected `None`.
___
*[`gsc.py` line 474 at r1](https://reviewable.io/reviews/gramineproject/gsc/244#-OTRQTf3A6EyVg-leYAB:-OTRQTf3A6EyVg-leYAC:b61ab1g) ([raw file](https://github.com/gramineproject/gsc/blob/6660bc9f9eafebdb328306af311012521ee75eae/gsc.py#L474)):*
> ```Python
> 
>     config = yaml.safe_load(args.config_file)
>     if 'Image' in config.get('Gramine'):
> ```

ditto (`TypeError`)


<!-- Sent from Reviewable.io -->

mkow reviewed (commented):
@DukeDavis12: Please use Reviewable for the review and reply to my comments.

> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/244#-:-OTb0aZH8ubF-Xz91E3-:b-nanqc3)* status: 0 of 1 files reviewed, 5 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @DukeDavis12)



<!-- Sent from Reviewable.io -->

DukeDavis12 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/244)* status: 0 of 1 files reviewed, 5 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @mkow)

___
*[`-- commits` line 5 at r1](https://reviewable.io/reviews/gramineproject/gsc/244#-OTRPtPbAH5AWteSiXfM:-OTaycRQ9de5nPI7SOvj:bb6un1j):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

nitpick
</blockquote></details>

Done
___
*[`-- commits` line 5 at r1](https://reviewable.io/reviews/gramineproject/gsc/244#-OTRQ--M991aysMcy8Mk:-OTayfBQ6G5QsGg1Fpof:b-896fix):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Also? Isn't that the only thing it prevents?
</blockquote></details>

Done.
___
*[`gsc.py` line 352 at r1](https://reviewable.io/reviews/gramineproject/gsc/244#-OTRQMjC0RYAr_zjgO_d:-OTaykOUCX6tZK0cTSHt:b-824ljb) ([raw file](https://github.com/gramineproject/gsc/blob/6660bc9f9eafebdb328306af311012521ee75eae/gsc.py#L352)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

This will still fail if `Gramine` key is not present in the config (you'll get a `TypeError`).
</blockquote></details>

Now we are checking whether 'Gramine' key is available. If it is not available we are printing Missing `Gramine` key in config file.


<!-- Sent from Reviewable.io -->

DukeDavis12 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/244)* status: 0 of 1 files reviewed, 5 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @mkow)

___
*[`gsc.py` line 353 at r1](https://reviewable.io/reviews/gramineproject/gsc/244#-OTRQeOY5bZGo65rMCTs:-OTbEHku4hkCgT6zVnaN:bppe4k5) ([raw file](https://github.com/gramineproject/gsc/blob/6660bc9f9eafebdb328306af311012521ee75eae/gsc.py#L353)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

This still won't work if the name is missing, `get_docker_image` will get `None` instead of the name and fail.
We either have to provide a default value (which doesn't make sense for this config option, I think), or just error out on the key access (i.e. here) instead of somewhere deeper in the logic. It will be much clearer to the user if they get `"Image" key not found` error, instead of something saying that the Docker daemon got an unexpected `None`.
</blockquote></details>

Now we check for following cases:
Check if Gramine is present
     Check if [Gramine][Image] is present or not 
     if not we check whether [Gramine][Repository] or [Gramine][Branch]
     else we print the error.
___
*[`gsc.py` line 474 at r1](https://reviewable.io/reviews/gramineproject/gsc/244#-OTRQTf3A6EyVg-leYAB:-OTbEvQy1njSV4axcWz7:b-896fix) ([raw file](https://github.com/gramineproject/gsc/blob/6660bc9f9eafebdb328306af311012521ee75eae/gsc.py#L474)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

ditto (`TypeError`)
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

DukeDavis12 reviewed (commented):
Done

> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/244#-:-OTbEzue5EEXvKLkAgDW:bb6un1j)* status: 0 of 1 files reviewed, 5 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @mkow)



<!-- Sent from Reviewable.io -->

donporter reviewed (changes requested):
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/244)* status: all files reviewed, 8 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @mkow)

___
*[`gsc.py` line 353 at r1](https://reviewable.io/reviews/gramineproject/gsc/244#-OTRQeOY5bZGo65rMCTs:-OTdiSw54fTklq725KQF:b-na1xiv) ([raw file](https://github.com/gramineproject/gsc/blob/6660bc9f9eafebdb328306af311012521ee75eae/gsc.py#L353)):*
<details><summary><i>Previously, DukeDavis12 (Davis Benny) wrote…</i></summary><blockquote>

Now we check for following cases:
Gramine section is present.
if Gramine is present
     Check if [Gramine][Image] is present or not 
     if not we check whether [Gramine][Repository] or [Gramine][Branch]
     else we print the error.
</blockquote></details>

Nit: Can you save the result of the first config.get?
___
*[`gsc.py` line 474 at r1](https://reviewable.io/reviews/gramineproject/gsc/244#-OTRQTf3A6EyVg-leYAB:-OTdii8b6NArzb1kVJBv:bduvd6v) ([raw file](https://github.com/gramineproject/gsc/blob/6660bc9f9eafebdb328306af311012521ee75eae/gsc.py#L474)):*
<details><summary><i>Previously, DukeDavis12 (Davis Benny) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Nit: Reuse result of first config.get()
___
*[`gsc.py` line 354 at r2](https://reviewable.io/reviews/gramineproject/gsc/244#-OTdiZHl0x2ox1rQgtsB:-OTdiZHl0x2ox1rQgtsC:b-89a4l7) ([raw file](https://github.com/gramineproject/gsc/blob/f394b6078dfd4e08159499b435b417d2d3b47246/gsc.py#L354)):*
> ```Python
>     if config.get('Gramine') :
>         if config.get('Gramine').get('Image'):
>             gramine_image_name = config.get('Gramine', {}).get('Image')
> ```

Ditto
___
*[`gsc.py` line 358 at r2](https://reviewable.io/reviews/gramineproject/gsc/244#-OTdia2i5Fy1LgCxGUDi:-OTdia2i5Fy1LgCxGUDj:b-1n3wsk) ([raw file](https://github.com/gramineproject/gsc/blob/f394b6078dfd4e08159499b435b417d2d3b47246/gsc.py#L358)):*
> ```Python
>                 print(f'Cannot find base-Gramine Docker image `{gramine_image_name}`.')
>                 sys.exit(1)
>         elif ((config.get('Gramine').get('Repository', None) is None)
> ```

And ditto
___
*[`gsc.py` line 359 at r2](https://reviewable.io/reviews/gramineproject/gsc/244#-OTdibY26aUXOnakyEe8:-OTdibY26aUXOnakyEe9:b-7ropln) ([raw file](https://github.com/gramineproject/gsc/blob/f394b6078dfd4e08159499b435b417d2d3b47246/gsc.py#L359)):*
> ```Python
>                 sys.exit(1)
>         elif ((config.get('Gramine').get('Repository', None) is None)
>               or (config.get('Gramine').get('Branch', None) is None)):
> ```

ditto


<!-- Sent from Reviewable.io -->

DukeDavis12 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/244)* status: 0 of 1 files reviewed, 8 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @donporter and @mkow)

___
*[`gsc.py` line 353 at r1](https://reviewable.io/reviews/gramineproject/gsc/244#-OTRQeOY5bZGo65rMCTs:-OTeDEUG7oZ3SUVEwsum:bb6un1j) ([raw file](https://github.com/gramineproject/gsc/blob/6660bc9f9eafebdb328306af311012521ee75eae/gsc.py#L353)):*
<details><summary><i>Previously, donporter (Don Porter) wrote…</i></summary><blockquote>

Nit: Can you save the result of the first config.get?
</blockquote></details>

Done
___
*[`gsc.py` line 474 at r1](https://reviewable.io/reviews/gramineproject/gsc/244#-OTRQTf3A6EyVg-leYAB:-OTeDJfR00KFeK8Kuo8A:b-896fix) ([raw file](https://github.com/gramineproject/gsc/blob/6660bc9f9eafebdb328306af311012521ee75eae/gsc.py#L474)):*
<details><summary><i>Previously, donporter (Don Porter) wrote…</i></summary><blockquote>

Nit: Reuse result of first config.get()
</blockquote></details>

Done.
___
*[`gsc.py` line 354 at r2](https://reviewable.io/reviews/gramineproject/gsc/244#-OTdiZHl0x2ox1rQgtsB:-OTeDFp28S-o9tk0nM3p:b-896fix) ([raw file](https://github.com/gramineproject/gsc/blob/f394b6078dfd4e08159499b435b417d2d3b47246/gsc.py#L354)):*
<details><summary><i>Previously, donporter (Don Porter) wrote…</i></summary><blockquote>

Ditto
</blockquote></details>

Done.
___
*[`gsc.py` line 358 at r2](https://reviewable.io/reviews/gramineproject/gsc/244#-OTdia2i5Fy1LgCxGUDi:-OTeDOMj9Mk6JgVzvPtP:b-896fix) ([raw file](https://github.com/gramineproject/gsc/blob/f394b6078dfd4e08159499b435b417d2d3b47246/gsc.py#L358)):*
<details><summary><i>Previously, donporter (Don Porter) wrote…</i></summary><blockquote>

And ditto
</blockquote></details>

Done.
___
*[`gsc.py` line 359 at r2](https://reviewable.io/reviews/gramineproject/gsc/244#-OTdibY26aUXOnakyEe8:-OTeDIONCTBySLlzdRly:b-896fix) ([raw file](https://github.com/gramineproject/gsc/blob/f394b6078dfd4e08159499b435b417d2d3b47246/gsc.py#L359)):*
<details><summary><i>Previously, donporter (Don Porter) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mkow reviewed (changes requested):
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/244)* status: all files reviewed, 8 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @donporter)

___
*[`gsc.py` line 359 at r3](https://reviewable.io/reviews/gramineproject/gsc/244#-OVyX_wbDnszokWPxgtN:-OVyX_wbDnszokWPxgtO:bdplrks) ([raw file](https://github.com/gramineproject/gsc/blob/64ca6ed7c0e71a611b42116cdbd55e0e2ad9ea51/gsc.py#L359)):*
> ```Python
>                 print(f'Cannot find base-Gramine Docker image `{gramine_image_name}`.')
>                 sys.exit(1)
>         elif ((gramine_config.get('Repository') is None)
> ```

ditto after `or`



_Suggestion:_
```Python
elif ('Repository' not in gramine_config
```
___
*[`gsc.py` line 360 at r3](https://reviewable.io/reviews/gramineproject/gsc/244#-OVyXEp3BxL7_YVE5rIC:-OVyXEp3BxL7_YVE5rID:undefined) ([raw file](https://github.com/gramineproject/gsc/blob/64ca6ed7c0e71a611b42116cdbd55e0e2ad9ea51/gsc.py#L360)):*
> ```Python
>                 sys.exit(1)
>         elif ((gramine_config.get('Repository') is None)
>               or (gramine_config.get('Branch') is None)):
> ```





_Suggestion:_
```Python
        elif ('Repository' not in gramine_config
             or 'Branch' not in gramine_config):
```
___
*[`gsc.py` line 361 at r3](https://reviewable.io/reviews/gramineproject/gsc/244#-OVyWpOWE47XXu0z2O3w:-OVyWpOWE47XXu0z2O3x:bicusm2) ([raw file](https://github.com/gramineproject/gsc/blob/64ca6ed7c0e71a611b42116cdbd55e0e2ad9ea51/gsc.py#L361)):*
> ```Python
>         elif ((gramine_config.get('Repository') is None)
>               or (gramine_config.get('Branch') is None)):
>             print('Gramine.Image or Gramine.Repository or Gramine.Branch key not found')
> ```

Please split this into separate errors for better user experience (consider that before this PR the user would know which key exactly was missing, now it's coalesced).
___
*[`gsc.py` line 365 at r3](https://reviewable.io/reviews/gramineproject/gsc/244#-OVyWVOPD4dwKAS5zlaI:-OVyWVOPD4dwKAS5zlaJ:ba6blsh) ([raw file](https://github.com/gramineproject/gsc/blob/64ca6ed7c0e71a611b42116cdbd55e0e2ad9ea51/gsc.py#L365)):*
> ```Python
>     else:
>         print('Error: Missing `Gramine` key in config file.')
>         sys.exit(1)
> ```

Please restructure this whole check. Instead of nesting it's clearer to check the prerequisites first, and then do the whole logic (i.e. "fail fast/early", instead of moving the fail cases to the `else` branches and nesting successes).
___
*[`gsc.py` line 485 at r3](https://reviewable.io/reviews/gramineproject/gsc/244#-OVyXtvm6tPXygMVjCi9:-OVyXtvn3MDFc_GZswSJ:undefined) ([raw file](https://github.com/gramineproject/gsc/blob/64ca6ed7c0e71a611b42116cdbd55e0e2ad9ea51/gsc.py#L485)):*
> ```Python
>     gramine_config = config.get('Gramine')
>     if gramine_config:
>         if gramine_config.get('Image'):
> ```





_Suggestion:_
```Python
    if 'Image' in config.get('Gramine', {}):
```


<!-- Sent from Reviewable.io -->

kailun-qin reviewed (commented):
> @kailun-qin made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/244)* status: all files reviewed, 8 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on donporter and DukeDavis12).

___
*[a discussion](https://reviewable.io/reviews/gramineproject/gsc/244#-Ooju_Sq98WifSyMiNjZ:-Ooju_Sq98WifSyMiNj_:b-63asw4) (no related file):*
@DukeDavis12: Hi Davis, the issue has blocked several internal and external users, and I'd like to get it closed. Are you going to continue working on this PR and address the remaining comments? Otherwise, I can also help take over and finish it. Thanks!


<!-- Sent from Reviewable.io -->

donporter reviewed (dismissed):
> @donporter reviewed 1 file and all commit messages, and resolved 3 discussions.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/244)* status: all files reviewed, 5 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: OSCAR Lab) (waiting on DukeDavis12).



<!-- Sent from Reviewable.io -->

kailun-qin committed:
Use config.get() for safer access to configuration

This commit updates direct dictionary access of configuration with
safer config.get(). This change prevents potential `KeyError`
exceptions if the keys are missing.

Co-authored-by: Davis Benny <davis.benny@intel.com>
Signed-off-by: Davis Benny <davis.benny@intel.com>
Signed-off-by: Kailun Qin <kailun.qin@intel.com>

kailun-qin reviewed (commented):
> @kailun-qin reviewed 1 file and all commit messages, and made 6 comments.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/244)* status: all files reviewed, 5 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on mkow).

___
*[a discussion](https://reviewable.io/reviews/gramineproject/gsc/244#-Ooju_Sq98WifSyMiNjZ:-Op79dNpEwGxcx-jpnBn:bn3x2oa) (no related file):*
<details><summary><i>Previously, kailun-qin (Kailun Qin) wrote…</i></summary><blockquote>

@DukeDavis12: Hi Davis, the issue has blocked several internal and external users, and I'd like to get it closed. Are you going to continue working on this PR and address the remaining comments? Otherwise, I can also help take over and finish it. Thanks!
</blockquote></details>

I'm taking over this PR from Davis, as he's moved on to another project (see https://github.com/gramineproject/gsc/pull/238#issuecomment-4167348084).
___
*[`gsc.py` line 359 at r3](https://reviewable.io/reviews/gramineproject/gsc/244#-OVyX_wbDnszokWPxgtN:-Op79zvH8uEspHe1kHqm:b-896fix) ([raw file](https://github.com/gramineproject/gsc/blob/a8872624823673e1399422e6f342fed6ba4b6e1a/gsc.py#L359)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

ditto after `or`
</blockquote></details>

Done.
___
*[`gsc.py` line 361 at r3](https://reviewable.io/reviews/gramineproject/gsc/244#-OVyWpOWE47XXu0z2O3w:-Op7A0lR9yOZQzH0Bvjq:b-896fix) ([raw file](https://github.com/gramineproject/gsc/blob/a8872624823673e1399422e6f342fed6ba4b6e1a/gsc.py#L361)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Please split this into separate errors for better user experience (consider that before this PR the user would know which key exactly was missing, now it's coalesced).
</blockquote></details>

Done.
___
*[`gsc.py` line 365 at r3](https://reviewable.io/reviews/gramineproject/gsc/244#-OVyWVOPD4dwKAS5zlaI:-Op7A1rM3TWM1M7hh7EE:b-896fix) ([raw file](https://github.com/gramineproject/gsc/blob/a8872624823673e1399422e6f342fed6ba4b6e1a/gsc.py#L365)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Please restructure this whole check. Instead of nesting it's clearer to check the prerequisites first, and then do the whole logic (i.e. "fail fast/early", instead of moving the fail cases to the `else` branches and nesting successes).
</blockquote></details>

Done.
___
*[`gsc.py` line 360 at r3](https://reviewable.io/reviews/gramineproject/gsc/244#-OVyXEp3BxL7_YVE5rIC:-Op7A-u36ZFhzH5n0W-N:b-896fix) ([raw file](https://github.com/gramineproject/gsc/blob/a8872624823673e1399422e6f342fed6ba4b6e1a/gsc.py#L360)):*
> ```Python
>                 sys.exit(1)
>         elif ((gramine_config.get('Repository') is None)
>               or (gramine_config.get('Branch') is None)):
> ```

Done.
___
*[`gsc.py` line 485 at r3](https://reviewable.io/reviews/gramineproject/gsc/244#-OVyXtvm6tPXygMVjCi9:-Op7A7stEziT3S0EIRGE:b-896fix) ([raw file](https://github.com/gramineproject/gsc/blob/a8872624823673e1399422e6f342fed6ba4b6e1a/gsc.py#L485)):*
> ```Python
>     gramine_config = config.get('Gramine')
>     if gramine_config:
>         if gramine_config.get('Image'):
> ```

Done.


<!-- Sent from Reviewable.io -->

donporter reviewed (approved):
> @donporter reviewed 1 file and all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/244)* status: all files reviewed, 5 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: OSCAR Lab) (waiting on mkow).



<!-- Sent from Reviewable.io -->

donporter reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/244)* status: all files reviewed, 5 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: OSCAR Lab) (waiting on mkow).



<!-- Sent from Reviewable.io -->

mkow reviewed (approved):
> @mkow reviewed 1 file and all commit messages, and resolved 5 discussions.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/244)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved.



<!-- Sent from Reviewable.io -->
